간단한 CRUD API 구현

sandropark opened the pull request:
# 작업사항
- 기본적인 CRUD API 추가
- testcontainer로 통합테스트 작성 (Fixture Monkey 써봤습니다.)
- github actions로 PR 등록 시 자동 빌드 설정

# 참고사항
#4 

# Review Rule
P1: 꼭 반영해주세요 (Request changes)
P2: 적극적으로 고려해주세요 (Request changes)
P3: 웬만하면 반영해 주세요 (Comment)
P4: 반영해도 좋고 넘어가도 좋습니다 (Approve)
P5: 그냥 사소한 의견입니다 (Approve)

sandropark committed:
feat: ktlint 제거

sandropark committed:
feat: User CRUD 기본틀 구현

sandropark committed:
refactor: DDD 아키텍처 적용 및 예외 처리 시스템 개선

- 패키지 구조를 DDD(Domain-Driven Design) 형태로 재구성
  * 계층형 아키텍처에서 도메인 중심 구조로 변경
  * user 도메인을 domain/application/presentation 계층으로 분리
  * shared 패키지로 공통 컴포넌트 분리

- 예외 처리 시스템 통합 및 개선
  * 개별 예외 클래스를 통합 예외 시스템으로 변경
  * ErrorCode enum으로 에러 코드 중앙 관리
  * RealtimeAppException으로 비즈니스 예외 통합 처리

- DTO to Entity 변환 로직 개선
  * UserCreateRequest에 toEntity() 메서드 추가
  * User 엔티티에 update() 메서드 추가

- 모든 테스트 통과 확인 (9/9)

sandropark committed:
feat: 이메일 중복 불가 로직 추가

sandropark committed:
refactor: 엔티티의 auditing 필드는 공통 클래스로 빼기

sandropark committed:
test: 통합테스트 클래스 이름 명확히 수정 및 경로 변경

- UserControllerIntegrationTest → UserApiIntegrationTest
- 기존 controller 관련 경로에서 user 패키지로 이동
- todo.md 업데이트 (작업 완료 표시)

sandropark committed:
test: 통합테스트에 DisplayName 추가

- 테스트 케이스별로 DisplayName 명시하여 가독성 및 의도 명확화
- UserApiIntegrationTest 클래스의 모든 테스트 메서드에 적용

sandropark committed:
refactor: User API 엔드포인트 URL 버전링 추가

sandropark committed:
refactor: 예외 클래스 구조 변경

- ErrorCode가 필요없기 때문에 삭제하고 Exception 내에서 상황에 맞는 팩터리 메서드를 정의

sandropark committed:
refactor: build.gradle 코틀린으로 변경

sandropark committed:
refactor: 코드 가독성 개선 및 불필요한 로직 제거

- UserService, UserController 메서드 리팩토링
- GlobalExceptionHandler의 ValidationException 처리 로직 간결화
- 테스트 코드 전반 리팩토링 및 불필요한 MockMvc 코드 제거

sandropark committed:
test: Fixture Monkey 추가

-  검증문 테스트 제거

sandropark committed:
feat: User API Swagger 문서화 추가

- SpringDoc OpenAPI 의존성 추가
- Swagger 설정 클래스 구현
- User DTO에 Schema 어노테이션 추가
- UserController에 API 문서화 어노테이션 추가
- todo.md 업데이트

sandropark committed:
config: 환경별 설정 파일 분

- 개발/운영 환경별 application.yml 파일 생성
- 공통 설정과 환경별 설정 분리
- 데이터베이스 설정을 환경별로 구분

sandropark committed:
feat: GitHub PR 자동화 설정 추가

- CODEOWNERS 파일로 realtime-team 자동 리뷰어 지정
- PR 템플릿으로 표준화된 PR 형식 제공

DevRunner21 commented on src/main/kotlin/com/sandro/realtime/shared/domain/BaseAuditEntity.kt:
```diff
@@ -0,0 +1,23 @@
+abstract class BaseAuditEntity(
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    open val id: Long? = null,
```
[P4] id를 상위 클래스로 추상화 하셨군요!
이렇게 하면 매 Entity마다 id 필드를 정의할 필요는 없을 것 같아요.

다만 id 필드가 다른 타입으로 바뀔 경우 이 클래스를 사용하지 못하게 될 것 같은데,
id 필드와 createdAt, updatedAt의 추상 클래스를 분리하는건 어떨까요?

sandropark replied:
그져 근데 경험상 id는 거의 auto increment로 사용해서 일단은 이렇게 했습니다. ㅎㅎ 나중에 내려도 큰 작업은 아니라 부담이 별로 없을 것 같기도 했구요~

DevRunner21 commented on src/main/kotlin/com/sandro/realtime/user/domain/model/User.kt:
```diff
@@ -0,0 +1,20 @@
+    var email: String,
+) : BaseAuditEntity() {
+
+    fun update(name: String) {
```
[P4] 실제로 변경되는 값은 name 밖에 없는데 메서드 네이밍이 그냥 'update'라서 
엔티티 전체 필드를 변경할 수 있는 메서드로 오해할 수 있을 것 같아요.

필드가 더 추가되면서 용도가 더 범용적으로 사용되도록 수정하실 수도 있지만 ㅎㅎ
만약 name 수정만을 위한 메서드라면 updateName 같은 네이밍은 어떨까요??

sandropark replied:
여러 필드를 가정하고 만든거라 ㅎㅎ

DevRunner21 commented on src/test/kotlin/com/sandro/realtime/user/UserApiIntegrationTest.kt:
```diff
@@ -0,0 +1,222 @@
+    @Autowired
+    private lateinit var userRepository: UserRepository
+
+    private val fixtureMonkey = FixtureMonkey.builder()
```
FixtureMonkey!!  👍👍

DevRunner21 commented on src/test/kotlin/com/sandro/realtime/user/UserApiIntegrationTest.kt:
```diff
@@ -0,0 +1,222 @@
+    @Test
+    @DisplayName("유저 조회 - 전체 유저 목록을 정상적으로 조회해야 한다")
+    fun `should get all users successfully`() {
+        // given
+        val userRequests = listOf(
+            createValidUserRequest(),
+            fixtureMonkey.giveMeBuilder<UserCreateRequest>()
+                .setExp(UserCreateRequest::name, "김철수")
+                .setExp(UserCreateRequest::email, "kim@example.com")
+                .sample()
+        )
+        val users = userRequests.map { it.toEntity() }
+        userRepository.saveAll(users)
```
[P4] FixtureMonkey로 가데이터를 세팅해주는 부분이네요!
확실히 툴을 쓰니까 간단해지는군요 ㅋㅋ

다만 여기서 Request를 통해서 Entity를 만들어서 저장해주시는 걸로 보이는데,
이렇게 되면 컨트롤러나 서비스 로직을 거치지 않으면서도 DTO 변환 코드가 테스트에 끼어드는 상황이 되버려서 DTO 로직이 바뀌었을때 가데이터 세팅부터 깨지는 상황이 발생할 수 있을 듯 합니다.

그래서 저같은 경우 가데이터 세팅은 Entity를 직접 저장하는 방식을 선호했던 것 같아요!
FixtureMonkey를 사용해서 Entity를 직접 만들수도 있을듯 한데 그런 방식은 어떨까요??

sandropark replied:
원래는 엔티티 직접 만드는 방식이었는데 LLM이 이렇게 수정해놨네요 ㅎㅎ
저는 단순히 필드만 매핑하는 변환 로직은 DTO안에 두는 걸 좋아해서 이렇게 했습니다. 서비스에서 해야하는 게 많으면 서비스 단에 두고요 ㅎㅎ
여기는 그냥 엔티티를 만드는 쪽이 더 깔금할 것 같네요. 수정하겠습니다! 👍

DevRunner21 commented on src/test/kotlin/com/sandro/realtime/user/UserApiIntegrationTest.kt:
```diff
@@ -0,0 +1,222 @@
+        val userRequest = createValidUserRequest()
+        val savedUser = userRepository.save(userRequest.toEntity())
+        val userId = savedUser.id
+        val updateRequest = fixtureMonkey.giveMeBuilder<UserUpdateRequest>()
+            .setExp(UserUpdateRequest::name, "홍길동수정")
+            .sample()
```
[P5] FixtureMonkey는  다양한 랜덤 값을 빠르게 만들어주는 툴로만 알고 있었는데,
이렇게 DTO 객체도 생성해줄 수 있군요!

다만 이 부분은  FixtureMonkey를 사용하지 않고 그냥 생성해도 될 듯한데,
FixtureMonkey로 DTO를 생성하신 이유가 있을까요??
( 저도 테스트 시 도입해보려고 해서 궁금한게 많네요 ㅜㅜ )

sandropark replied:
마자요 ㅋㅋ 이 부분은 FixtureMonkey를 사용하는 이점이 없긴하죠. Claude로 그냥 만들라고 시키니까 이렇게 해뒀네요 ㅎㅎ

DevRunner21 commented on src/main/kotlin/com/sandro/realtime/user/domain/model/User.kt:
```diff
@@ -0,0 +1,20 @@
+@Table(name = "users")
+class User(
+    @Column(nullable = false)
+    var name: String,
```
[P5] 이건 좀 과한 설계일지는 모르지만
예를들어 name 필드에 DB 스키마에 있는 제한사항보다 긴 문자열이 들어오면 어떻게 되나요??

예전에 DDD관련 책에서 Entity 정합성 체크를 빡쎼게 관리하던 코드가 생각나서 ㅋㅋ
산들님은 이런 부분들은 어떻게 생각하시는지 궁금합니다!

sandropark replied:
validation에 길이 검증 로직을 추가하는 게 좋을 것 같네요 ㅎㅎ
사실 엔티티에 빡센 검증 로직이 들어가는 게 좋을 것 같기도 한데... Spring vaiidation으로 controller에서 튕구고 싶기도 하고.. 빡세게 하려면 둘 다 넣어도 좋을 것 같네요..! 저는 그정도는 아니라 앞에서 잘 걸러질 거라 가정하렵니다!
섬세하시군요 ㅎㅎ

DevRunner21 reviewed (commented):
리뷰 드렸습니다! 👍👍

sandropark committed:
feat: 사용자 유효성 검증 강화 및 테스트 코드 개선

- name/email 필드 길이 제한 추가 (@Size 어노테이션)
- Fixture Monkey 1.1.14 업그레이드
- Spring validation을 이용해서 유효한 객체가 생성되도록 설정

ebcode2021 commented on src/test/kotlin/com/sandro/realtime/user/UserApiIntegrationTest.kt:
```diff
@@ -74,10 +80,8 @@ class UserApiIntegrationTest {
     @DisplayName("유저 생성 - 중복된 이메일로 생성 시 409 에러가 발생해야 한다")
```
409 Conflict는 실제로 처음보네요~.~

sandropark replied:
저도 맨날 400만 쓰다가 claude가 추천해줘서 써봤습니다..!

ebcode2021 reviewed (approved):
테스트코드 수정된 부분을 봤는데, 세심하게 잘 작성하신 것 같습니다.
fixtureMonkey 라는 것을 이번에 처음 알게되었는데, 저도 한 번 써보겠습니다!

sandropark commented:
> 테스트코드 수정된 부분을 봤는데, 세심하게 잘 작성하신 것 같습니다. fixtureMonkey 라는 것을 이번에 처음 알게되었는데, 저도 한 번 써보겠습니다!

저번에 흥래님이 언급하셔서 한 번 써봤습니다! 일하면서 테스트 짤 때 input 데이터 만드는데 힘들었어서 써봤는데 좋은 것 같아요!! 회사에서 슬쩍 사용해봐야겠어요 ㅎㅎ
